homeassistant/components/emoncms/sensor.py

```diff
@@ -57,7 +58,8 @@
             {cv.positive_int: vol.All(cv.string, vol.Length(min=1))}
         ),
         vol.Optional(CONF_VALUE_TEMPLATE): cv.template,
-        vol.Optional(CONF_UNIT_OF_MEASUREMENT, default=DEFAULT_UNIT): cv.string,
+        vol.Optional(CONF_UNIT_OF_MEASUREMENT,default=DEFAULT_UNIT): cv.string,
+        vol.Optional(CONF_USE_EMONCMS_UNIT,default=DEFAULT_CONF_USE_EMONCMS_UNIT): cv.boolean,
```

Comment:
Tbh I would prefer a breaking change instead of introducing such an option. An integration should always use the unit of measurement from the api if it is available. It definitely would be more consistent with other integrations and I cannot think of another one right now, where such an option is used.

Reply:
It needs to be your design choice as you have not really liked either of my options.  Because of the DEFAULT_VALUE, it makes it very difficult to migrate cleanly.

Are you saying you want 
1. the option removed
2. The API unit used if present
3. the UNIT_OF_MEASUREMENT that has either been set of where the default has picked it up?

The code would then become...
```
unit_of_measurement = elem.get("unit",config.get(CONF_UNIT_OF_MEASUREMENT))
```

Reply:
1. Yes, `CONF_USE_EMONCMS_UNIT` should be removed.
2. Yes, exactly.

> I have missed a use case here, the 'unit' key may exist in the API call but it may not be set in Emoncms so will be just an empty string.

Can the user set the unit in Emoncms manually? Because if that is the case I would even remove the unit of measurement and add a note to the documentation that the user should add the unit in Emoncms manually.

Reply:
Any updates on this? 🙂

Reply:
Just in. The problem this was added to the API at some point so older instances won't work (and upgrading is not always straight forward so not always an option).

Just trying to check and see when it was added.

For now, I'd rather keep it but mark it as a feature that will go at some point in the future. I've seen that with other integrations - not sure how that message would be added.

Thanks for your patience! 😁

Reply:
Okay thanks for looking deepr into this :)

IMHO we should leave your PR as it is, to have the following logic:

1. take the unit from the API if it is available
2. otherwise take the unit from config, either specified or the default one

Consequently, all sensors will have a unit of measurement to make it easier to display them in graphs etc. and the new feature is included.
However, I would suggest to add a breaking change and add a note to the documentation describing the logic and how to upgrade to new firmware or how to set this unit in emoncms.

What do you think about all of this?

Reply:
Ok great. I'll amend the PR to show it as a breaking change and update the documentation (again) 😁.

I have also worked out when the feature was added (summer 2018) so I can add that as well.